src/modules/python/pbs_hooks/PBS_power.PY

```diff
@@ -0,0 +1,191 @@
+        event.reject(str(e))
+    return power
+
+
+def vnodes_enabled(job):
+    # see if power operations are allowed on all job vnodes
+    for vn in _get_vnodes(job):
+        if not _svr_vnode(vn).power_provisioning:
+            pbs.logjobmsg(job.id,
+                "power functionality is disabled on vnode %s" % vn)
+            return False
+    return True
+
+
+# Accept if event not serviceable.
+this_event = pbs.event()
+if this_event.type not in [pbs.EXECJOB_PROLOGUE, pbs.EXECJOB_EPILOGUE,
+                           pbs.EXECJOB_BEGIN, pbs.EXECJOB_END,
+                           pbs.EXECHOST_STARTUP, pbs.EXECHOST_PERIODIC]:
+    pbs.logmsg(pbs.LOG_WARNING,
+        "Event not serviceable for power provisioning.")
+    this_event.accept()
+
+
+# Set eoe values for my node
+if this_event.type == pbs.EXECHOST_STARTUP:
+    from pbs.v1._pmi_utils import _is_node_provisionable
+
+    # Don't connect if the server or sched is running.
+    if not _is_node_provisionable():
+        pbs.logmsg(pbs.LOG_DEBUG,
+            "Provisioning cannot be enabled on this host")
+        this_event.accept()
+    power = init_power(this_event)
+    profiles = power.query(pbs.Power.QUERY_PROFILE)
+    if profiles is not None:
+        me = pbs.get_local_nodename()
+        this_event.vnode_list[me].resources_available[
+            "eoe"] = power._map_profile_names(profiles)
+    power.disconnect()
+    this_event.accept()
+
+
+# Gather energy usage for all jobs
+if this_event.type == pbs.EXECHOST_PERIODIC:
+    # Check if any jobs are running
+    if len(this_event.job_list) == 0:
+        this_event.accept()
+
+    power = init_power(this_event)
+    for jobid in this_event.job_list.keys():
+        # set energy usage
+        job = this_event.job_list[jobid]
+        if not job.in_ms_mom():		# skip any jobs that MOM is not MS
+            continue
+        if not vnodes_enabled(job):	# skip if vnodes have power_provisioning=0
+            continue
+
+        try:
+            usage = power.get_usage(job)
+            if usage is not None:
+                job.resources_used["energy"] = usage
+        except Exception, e:
+            pbs.logmsg(pbs.LOG_ERROR, str(e))
+    power.disconnect()
+    this_event.accept()
+
+# From this point on, the event will have a job.
+this_job = this_event.job
+
+# The following code is needed until spid 329611 is fixed.
+if this_event.type == pbs.EXECJOB_BEGIN:
+    me = pbs.get_local_nodename()
+    if not _svr_vnode(me).power_provisioning:
+        this_event.accept()
+    requested_profile = str(this_job.schedselect).partition(
+        'eoe=')[2].partition('+')[0].partition(':')[0]
+    if requested_profile != "":
+        try:
+            this_event.vnode_list[me].current_eoe = requested_profile
+        except:
+            pass
+    this_event.accept()
+if this_event.type == pbs.EXECJOB_END:
+    me = pbs.get_local_nodename()
+    try:
+        this_event.vnode_list[me].current_eoe = None
```

Comment:
Once we are done with a job we revert the node to its original state. That happens when you deactivate the profile. Current_eoe shows what profile the node currently set to. It makes sense to reset the attribute once we are done with the job.